vignettes/task_plans.Rmd

```diff
@@ -0,0 +1,353 @@
+
+```
+step2 <- create_task_step(
+  step_name = 'plot',
+  target_name = function(target_name, ...) {
+    sprintf("out/%s.png", target_name)
+  },
+  command = function(steps, ...) {
+    sprintf('visualize(\'%s\', target_name)', steps[['prepare']]$target_name)
+  }
+)
+```
+
+As illustrated above, referencing the previous step's output can be handy for multi-step tasks where each step depends on output from the previous step.
+
+
+#### Creating a task plan
+
+You can now create a task_plan that wraps the tasks and steps together. In the following call to `create_task_plan`, a couple of things happen. The main thing is that the steps you identified are organized into an R object that `scipiper` can use efficiently later. The other thing is that when the default argument `add_complete=TRUE` is accepted (e.g., by not specifying `add_complete` at all), a final step gets tacked onto the list. The final step generates an .ind file for each task once all other steps for that task have been completed (or, if `final_steps` is specified, once those specified steps have been completed).
+
+The main benefit of this `complete` step and the corresponding .ind file is speed: iteration over the complete task plan can be more efficient because the presence of an .ind file for a task can be used as a first-cut indication that the task doesn't need to be built, such that iteration effort can be more efficiently spent on other tasks. The usual way that `remake` determines which tasks need to be rebuilt can be quite slow for large task lists or large output files, so speed here is a good thing.
+
+One drawback of the `add_complete` system and corresponding .ind files is that the speed shortcut only works if the presence of these .ind files is a fairly accurate description of which tasks are up to date. If you build a complete task plan once, change an input, and intend to rebuild, then it's strongly recommended that you delete the .ind files before calling `loop_tasks` again. If you don't, nothing will break, but your build will be slower and will not include automatic retries.
+
+```{r plan}
+task_plan_1 <- create_task_plan(
+  task_config$id,
+  list(step1, step2, step3),
+  final_steps='report',
+  ind_dir='states/log')
+```
+
+You'll need the `task_plan` R object for two things: to create a remake .yml makefile and to run `loop_tasks`.
+
+
+#### Creating a task makefile
+
+Here's how to create a task makefile from a task plan:
+
+```{r makefile, results='markup'}
+create_task_makefile(
+  task_plan=task_plan_1,
+  makefile='task_plan_1.yml',
+  include=c(),
+  packages=c('ggplot2'),
+  sources=c('script.R'),
+  file_extensions=c('ind'),
+  ind_complete=TRUE)
```

Comment:
I think we'll need to drop `ind_complete` from the arg list above when we merge the `task_combiners` branch into `master`.

Reply:
We could drop it right now, in anticipation of that change; the only awkwardness would be that the above command would yield `message('ind_complete=NA is deprecated; use TRUE or FALSE')` which we'd have to ignore for another few weeks to months. I think dropping it right now would be better than not dropping it and forgetting until the vignette breaks.

Reply:
Seems almost as simple as deleting, but there is a section that talks a lot about this option. See the `### Indicator files options` section. I think we would need to change more than just that argument when task_combiners gets merged.

Reply:
OK, waiting sounds like a good plan :+1: